pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10876974?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10876974?buildTab=artifacts#/) on release-23.1 @ [7bfb0ce4cf4e5d3c6c51fd97ced445f4c05fbdc8](https://github.com/cockroachdb/cockroach/commits/7bfb0ce4cf4e5d3c6c51fd97ced445f4c05fbdc8):


```
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/NoMode/Types (0.15s)
=== RUN   TestExplainRedactDDL/Explain/NoMode/Verbose
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/NoMode/Verbose (0.13s)
=== RUN   TestExplainRedactDDL/Explain/Opt/Memo
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/Opt/Memo (0.17s)
=== RUN   TestExplainRedactDDL/Explain/Opt/Verbose
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/Opt/Verbose (0.17s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/NoMode/Catalog
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/NoMode/Catalog (0.22s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/NoMode/Viz
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/NoMode/Viz (0.23s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/Plan/Memo
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/Plan/Memo (0.29s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/Plan/Shape
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/Plan/Shape (0.27s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/Plan/Viz
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/Plan/Viz (0.24s)
=== RUN   TestExplainRedactDDL/Explain/NoMode/Shape
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/NoMode/Shape (0.13s)
=== RUN   TestExplainRedactDDL/Explain/Opt/Catalog
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/Opt/Catalog (0.18s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/NoMode/NoFlag
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/NoMode/NoFlag (0.17s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/Plan/Catalog
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/Plan/Catalog (0.23s)
=== RUN   TestExplainRedactDDL/Explain/Opt/Shape
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/Opt/Shape (0.16s)
=== RUN   TestExplainRedactDDL/Explain/Vec/Catalog
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/Explain/Vec/Catalog (0.14s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/NoMode/Types
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/NoMode/Types (0.22s)
=== RUN   TestExplainRedactDDL/ExplainAnalyze/NoMode/Verbose
    explain_test_util.go:97: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
            --- FAIL: TestExplainRedactDDL/ExplainAnalyze/NoMode/Verbose (0.27s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainRedactDDL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29664

Hmm, well, I've got a repro with these statements:

```sql
SET statement_timeout = '5s';
CREATE TABLE table2 (col2_0 BOOL NOT NULL, col2_1 REGCLASS, col2_2 BOX2D NOT NULL, col2_3 GEOGRAPHY NOT NULL, col2_4 GEOGRAPHY, INVERTED INDEX (col2_2 ASC, col2_0 DESC, col2_3 ASC) WHERE col2_0, INDEX (col2_2 DESC), FAMILY (col2_4), FAMILY (col2_2), FAMILY (col2_1, col2_3), FAMILY (col2_0)) WITH (sql_stats_automatic_collection_enabled = false);
UPDATE table2 SET col2_1 = 0:::OID WHERE true;
```

But this appears to be specific to the test server. It reproduces when run in a test:

```
# michae2@michae2-crl-mbp ~crdb % git diff
diff --git a/pkg/ccl/testccl/sqlccl/explain_test.go b/pkg/ccl/testccl/sqlccl/explain_test.go
index 27ace734c65..1f7a0cfb86c 100644
--- a/pkg/ccl/testccl/sqlccl/explain_test.go
+++ b/pkg/ccl/testccl/sqlccl/explain_test.go
@@ -11,6 +11,7 @@ package sqlccl
 import (
        "context"
        gosql "database/sql"
+       "fmt"
        "strings"
        "testing"

@@ -24,6 +25,36 @@ import (
        "github.com/cockroachdb/errors"
 )

+func Test106663Repro(t *testing.T) {
+       defer leaktest.AfterTest(t)()
+       defer log.Scope(t).Close(t)
+
+       ctx := context.Background()
+
+       params, _ := tests.CreateTestServerParams()
+       s, sqlDB, _ := serverutils.StartServer(t, params)
+       defer s.Stopper().Stop(ctx)
+       defer sqlDB.Close()
+
+       query := func(sql string) (*gosql.Rows, error) {
+               return sqlDB.QueryContext(ctx, sql)
+       }
+
+       stmts := []string{
+               "SET statement_timeout = '5s'",
+               "CREATE TABLE table2 (col2_0 BOOL NOT NULL, col2_1 REGCLASS, col2_2 BOX2D NOT NULL, col2_3 GEOGRAPHY NOT NULL, col2_4 GEOGRAPHY, INVERTED INDEX (col2_2 ASC, col2_0 DESC, col2_3 ASC) WHERE col2_0, INDEX (col2_2 DESC), FAMILY (col2_4), FAMILY (col2_2), FAMILY (col2_1, col2_3), FAMILY (col2_0)) WITH (sql_stats_automatic_collection_enabled = false)",
+               "UPDATE table2 SET col2_1 = 0:::OID WHERE true",
+       }
+
+       for _, stmt := range stmts {
+               fmt.Println(stmt + ";")
+               _, err := query(stmt)
+               if err != nil {
+                       t.Error(err)
+               }
+       }
+}
+
 // TestExplainRedactDDL tests that variants of EXPLAIN (REDACT) do not leak
 // PII. This is very similar to sql.TestExplainRedact but includes CREATE TABLE
 // and ALTER TABLE statements, which could include partitioning (hence this is

# michae2@michae2-crl-mbp ~crdb % ./dev test pkg/ccl/testccl/sqlccl --filter=Test106663Repro
$ bazel test pkg/ccl/testccl/sqlccl:all --test_env=GOTRACEBACK=all --test_filter=Test106663Repro --test_sharding_strategy=disabled --test_output errors
WARNING: The following configs were expanded more than once: [dev]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
INFO: Invocation ID: 89a6ceb2-881f-4119-8dce-e8d04f5e3cc3
WARNING: The following configs were expanded more than once: [dev]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
INFO: Analyzed target //pkg/ccl/testccl/sqlccl:sqlccl_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
FAIL: //pkg/ccl/testccl/sqlccl:sqlccl_test (see /private/var/tmp/_bazel_michae2/239414c2d9c0d0876d35ed2ff4c5fe74/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin-fastbuild/testlogs/pkg/ccl/testccl/sqlccl/sqlccl_test/test.log)
INFO: From Testing //pkg/ccl/testccl/sqlccl:sqlccl_test:
==================== Test output for //pkg/ccl/testccl/sqlccl:sqlccl_test:
initialized metamorphic constant "span-reuse-rate" with value 43
initialized metamorphic constant "parse-json-impl" with value 0
initialized metamorphic constant "coldata-batch-size" with value 9
initialized metamorphic constant "kv.rangefeed.use_dedicated_connection_class.enabled" with value true
initialized metamorphic constant "mvcc-max-iters-before-seek" with value 0
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 30892
initialized metamorphic constant "kv.expiration_leases_only.enabled" with value true
initialized metamorphic constant "kv-batch-size" with value 1
initialized metamorphic constant "datum-row-converter-batch-size" with value 1
initialized metamorphic constant "inverted-joiner-batch-size" with value 1
initialized metamorphic constant "zig-zag-joiner-batch-size" with value 1
initialized metamorphic constant "ColIndexJoin-using-streamer-batch-size" with value 8364401
initialized metamorphic constant "max-batch-size" with value 5618
initialized metamorphic constant "max-batch-byte-size" with value 4423721
initialized metamorphic constant "parallel-scan-result-threshold" with value 992
initialized metamorphic constant "resolve-membership-single-scan-enabled" with value false
initialized metamorphic constant "copy-batch-size" with value 21598
initialized metamorphic constant "copy-fast-path-enabled-default" with value false
initialized metamorphic constant "backup.write_metadata_with_external_ssts.enabled" with value false
initialized metamorphic constant "restore-worker-concurrency" with value 1
initialized metamorphic constant "changefeed.event_consumer_worker_queue_size" with value 10
initialized metamorphic constant "changefeed.fast_gzip.enabled" with value false
initialized metamorphic constant "changefeed.new_webhook_sink_enabled" with value true
initialized metamorphic constant "changefeed.new_pubsub_sink_enabled" with value true
initialized metamorphic constant "aws-reuse-kms" with value false
initialized metamorphic constant "span-reuse-rate" with value 53
initialized metamorphic constant "COCKROACH_ENABLE_HDR_HISTOGRAMS" with value true
initialized metamorphic constant "parse-json-impl" with value 1
initialized metamorphic constant "coldata-batch-size" with value 1191
initialized metamorphic constant "kv.rangefeed.use_dedicated_connection_class.enabled" with value true
initialized metamorphic constant "use-mux-rangefeed" with value false
initialized metamorphic constant "mvcc-value-disable-simple-encoding" with value true
initialized metamorphic constant "storage.value_blocks.enabled" with value false
initialized metamorphic constant "addsst-rewrite-concurrency" with value 3
initialized metamorphic constant "split-scans-right-for-stats-first" with value true
initialized metamorphic constant "reproposals-v2" with value false
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 421147
initialized metamorphic constant "default-batch-bytes-limit" with value 19119
initialized metamorphic constant "kv-batch-size" with value 1
initialized metamorphic constant "inverted-joiner-batch-size" with value 1
initialized metamorphic constant "zig-zag-joiner-batch-size" with value 1
initialized metamorphic constant "spilling-queue-initial-len" with value 14
initialized metamorphic constant "merge-joiner-groups-buffer" with value 2
initialized metamorphic constant "direct-scans-enabled" with value true
initialized metamorphic constant "max-batch-size" with value 7331
initialized metamorphic constant "max-batch-byte-size" with value 24110025
initialized metamorphic constant "parallel-scan-result-threshold" with value 8514
initialized metamorphic constant "resolve-membership-single-scan-enabled" with value false
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 3
initialized metamorphic constant "use-index-lookup-for-descriptors-in-database" with value false
initialized metamorphic constant "copy-fast-path-enabled-default" with value false
initialized metamorphic constant "backup-merge-file-buffer-size" with value 8931329
initialized metamorphic constant "write-metadata-sst" with value true
initialized metamorphic constant "backup.write_metadata_with_external_ssts.enabled" with value false
initialized metamorphic constant "export_request_verbose_tracing" with value true
initialized metamorphic constant "restore-worker-concurrency" with value 7
initialized metamorphic constant "changefeed.mux_rangefeed.enabled" with value false
initialized metamorphic constant "changefeed.fast_gzip.enabled" with value false
I230712 19:08:34.963398 1 (gostd) rand.go:249  [-] 1  random seed: -744821518445971633
SET statement_timeout = '5s';
CREATE TABLE table2 (col2_0 BOOL NOT NULL, col2_1 REGCLASS, col2_2 BOX2D NOT NULL, col2_3 GEOGRAPHY NOT NULL, col2_4 GEOGRAPHY, INVERTED INDEX (col2_2 ASC, col2_0 DESC, col2_3 ASC) WHERE col2_0, INDEX (col2_2 DESC), FAMILY (col2_4), FAMILY (col2_2), FAMILY (col2_1, col2_3), FAMILY (col2_0)) WITH (sql_stats_automatic_collection_enabled = false);
UPDATE table2 SET col2_1 = 0:::OID WHERE true;
test logs left over in: /Users/michae2/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/9e6e471a7f7e2f77d39f6d917f95e758/logTest106663Repro1826989994
--- FAIL: Test106663Repro (0.79s)
    test_log_scope.go:167: test logs captured to: /Users/michae2/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/9e6e471a7f7e2f77d39f6d917f95e758/logTest106663Repro1826989994
    test_log_scope.go:81: use -show-logs to present logs inline
    explain_test.go:53: pq: internal error: optimizer factory constructor call stack exceeded max depth of 10000
    explain_test.go:56: -- test log scope end --
FAIL
I230712 19:08:35.760950 1 (gostd) testmain.go:109  [-] 1  Test //pkg/ccl/testccl/sqlccl:sqlccl_test exited with error code 1
================================================================================
Target //pkg/ccl/testccl/sqlccl:sqlccl_test up-to-date:
  _bazel/bin/pkg/ccl/testccl/sqlccl/sqlccl_test_/sqlccl_test
INFO: Elapsed time: 1.633s, Critical Path: 1.32s
INFO: 2 processes: 1 internal, 1 darwin-sandbox.
//pkg/ccl/testccl/sqlccl:sqlccl_test                                     FAILED in 1.3s
  /private/var/tmp/_bazel_michae2/239414c2d9c0d0876d35ed2ff4c5fe74/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin-fastbuild/testlogs/pkg/ccl/testccl/sqlccl/sqlccl_test/test.log

Executed 1 out of 1 test: 1 fails locally.
INFO: Build Event Protocol files produced successfully.
INFO: Build completed, 1 test FAILED, 2 total actions
ERROR: exit status 3
```

But does not reproduce on `cockroach demo`:

```
# michae2@michae2-crl-mbp ~crdb % cdemo
#
# Welcome to the CockroachDB demo database!
#
# You are connected to a temporary, in-memory CockroachDB cluster of 1 node.
#
# You are connected to tenant "demoapp", but can connect to the system tenant with
# '\connect' and the SQL url printed via '\demo ls'.
#
# Telemetry disabled by configuration.
#
# Reminder: your changes to data stored in the demo session will not be saved!
#
# If you wish to access this demo cluster using another tool, you will need
# the following details:
#
#   - Connection parameters:
#      (webui)    http://127.0.0.1:8080/demologin?password=demo65584&username=demo
#      (cli)      cockroach sql --certs-dir=/Users/michae2/.cockroach-demo -u demo -d cluster:demoapp/defaultdb
#      (sql)      postgresql://demo:demo65584@127.0.0.1:26257/defaultdb?options=-ccluster%3Ddemoapp&sslmode=require&sslrootcert=%2FUsers%2Fmichae2%2F.cockroach-demo%2Fca.crt
#
#   - Username: "demo", password: "demo65584"
#   - Directory with certificate files (for certain SQL drivers/tools): /Users/michae2/.cockroach-demo
#
# You can enter \info to print these details again.
#
# Server version: CockroachDB CCL v23.2.0-alpha.00000000-dev (darwin amd64, built , go1.19.10) (same version as client)
# Cluster ID: c2cd5eea-0e3d-4f93-42c2-e88656bf4cd7
# Organization: Cockroach Demo
#
# Enter \? for a brief introduction.
#
demo@127.0.0.1:26257/demoapp/defaultdb> SET statement_timeout = '5s';
                                     -> CREATE TABLE table2 (col2_0 BOOL NOT NULL, col2_1 REGCLASS, col2_2 BOX2D NOT NULL, col2_3 GEOGRAPHY NOT NULL, col2_4 GEOGRAPHY, INVERTED INDEX (col2_2 ASC, col2_0 DESC, col2_3 ASC) WHERE col2_0, INDEX (col2_2 DESC), FAMILY (col2_4), FAMILY (col2_2), FAMILY (col2_1, col2_3), FAMILY (col2_0)) WITH (sql_stats_automatic_collection_enabled = false);
                                     -> UPDATE table2 SET col2_1 = 0:::OID WHERE true;
SET

Time: 0ms total (execution 0ms / network 0ms)

CREATE TABLE

Time: 4ms total (execution 4ms / network 0ms)

UPDATE 0

Time: 15ms total (execution 15ms / network 0ms)

demo@127.0.0.1:26257/demoapp/defaultdb>
```

--- a/pkg/sql/opt/norm/mutation_funcs.go
+++ b/pkg/sql/opt/norm/mutation_funcs.go
@@ -11,8 +11,11 @@
 package norm
 
 import (
+	"fmt"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/opt"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/memo"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
 )
 
 // SimplifiablePartialIndexProjectCols returns the set of projected partial
@@ -172,8 +175,11 @@ func multiUsePartialIndexCols(mp *memo.MutationPrivate) opt.ColSet {
 // projected column's expression simplified to false if the column exists in
 // simplifiableCols.
 func (c *CustomFuncs) SimplifyPartialIndexProjections(
-	projections memo.ProjectionsExpr, simplifiableCols opt.ColSet,
-) memo.ProjectionsExpr {
+	projections memo.ProjectionsExpr,
+	passthrough opt.ColSet,
+	simplifiableCols opt.ColSet,
+	private *memo.MutationPrivate,
+) (memo.ProjectionsExpr, *memo.MutationPrivate) {
 	simplified := make(memo.ProjectionsExpr, len(projections))
 	for i := range projections {
 		if col := projections[i].Col; simplifiableCols.Contains(col) {
@@ -182,5 +188,40 @@ func (c *CustomFuncs) SimplifyPartialIndexProjections(
 			simplified[i] = projections[i]
 		}
 	}
-	return simplified
+
+	// Any simplifiable partial index expressions that are currently passthrough
+	// columns must be changed to projected false expressions.
+	simplifiableCols.IntersectionWith(passthrough)
+	if simplifiableCols.Empty() {
+		return simplified, private
+	}
+
+	// Copy the MutationPrivate in order to change the partial index columns.
+	simplifiedPrivate := *private
+	simplifiedPrivate.PartialIndexPutCols = append(opt.OptionalColList{}, private.PartialIndexPutCols...)
+	simplifiedPrivate.PartialIndexDelCols = append(opt.OptionalColList{}, private.PartialIndexDelCols...)
+
+	ord := len(simplifiedPrivate.PartialIndexPutCols)
+	for i, col := range simplifiedPrivate.PartialIndexPutCols {
+		if simplifiableCols.Contains(col) {
+			name := fmt.Sprintf("partial_index_put%d", ord+1)
+			newPutCol := c.f.Metadata().AddColumn(name, types.Bool)
+			simplified = append(simplified, c.f.ConstructProjectionsItem(memo.FalseSingleton, newPutCol))
+			simplifiedPrivate.PartialIndexPutCols[i] = newPutCol
+			ord++
+		}
+	}
+
+	ord = len(simplifiedPrivate.PartialIndexDelCols)
+	for i, col := range simplifiedPrivate.PartialIndexDelCols {
+		if simplifiableCols.Contains(col) {
+			name := fmt.Sprintf("partial_index_del%d", ord+1)
+			newDelCol := c.f.Metadata().AddColumn(name, types.Bool)
+			simplified = append(simplified, c.f.ConstructProjectionsItem(memo.FalseSingleton, newDelCol))
+			simplifiedPrivate.PartialIndexDelCols[i] = newDelCol
+			ord++
+		}
+	}
+
+	return simplified, &simplifiedPrivate
 }

--- a/pkg/sql/opt/norm/rules/mutation.opt
+++ b/pkg/sql/opt/norm/rules/mutation.opt
@@ -26,13 +26,21 @@
 (Update
     (Project
         $input
-        (SimplifyPartialIndexProjections
-            $projections
-            $simplifiableCols
+        (Let
+            (
+                $simplifiedProjections
+                $simplifiedPrivate
+            ):(SimplifyPartialIndexProjections
+                $projections
+                $passthrough
+                $simplifiableCols
+                $mutationPrivate
+            )
+            $simplifiedProjections
         )
         $passthrough
     )
     $uniqueChecks
     $fkChecks
-    $mutationPrivate
+    $simplifiedPrivate
 )

--- a/pkg/sql/opt/norm/testdata/rules/mutation
+++ b/pkg/sql/opt/norm/testdata/rules/mutation
@@ -299,3 +299,39 @@ update t74385
       └── projections
            ├── c:10 IS NULL [as=partial_index_put1:14, outer=(10)]
            └── CAST(NULL AS INT8) [as=b_new:13]
+
+# Regression for #106663. Also simplify passthrough partial index put/del
+# columns to false when the indexed columns and columns referenced in predicates
+# are not changing.
+
+exec-ddl
+CREATE TABLE t106663 (a INT, b INT, c BOOL, INDEX (b) WHERE c, FAMILY (a, b), FAMILY (c))
+----
+
+norm expect=SimplifyPartialIndexProjections
+UPDATE t106663 SET a = 1 WHERE true
+----
+update t106663
+ ├── columns: <none>
+ ├── fetch columns: a:7 b:8 rowid:10
+ ├── update-mapping:
+ │    └── a_new:13 => a:1
+ ├── partial index put columns: partial_index_put2:14
+ ├── partial index del columns: partial_index_del2:15
+ ├── cardinality: [0 - 0]
+ ├── volatile, mutations
+ └── project
+      ├── columns: a_new:13!null partial_index_put2:14!null partial_index_del2:15!null a:7 b:8 rowid:10!null
+      ├── key: (10)
+      ├── fd: ()-->(13-15), (10)-->(7,8)
+      ├── scan t106663
+      │    ├── columns: a:7 b:8 rowid:10!null
+      │    ├── partial index predicates
+      │    │    └── t106663_b_idx: filters
+      │    │         └── c:9 [outer=(9), constraints=(/9: [/true - /true]; tight), fd=()-->(9)]
+      │    ├── key: (10)
+      │    └── fd: (10)-->(7,8)
+      └── projections
+           ├── 1 [as=a_new:13]
+           ├── false [as=partial_index_put2:14]
+           └── false [as=partial_index_del2:15]

